docs: update AWS Cluster Autoscaler pack (#1411)
spectrocloud/librarium: content/docs/06-integrations/00-aws-autoscaler.md

Reviewer: Something to confirm with Vish:

- What's the most current version? The UI doesn't display version 1.24. It displays 1.26. 

- Are versions 1.22 and 1.0 are deprecated. These show up in the UI. If deprecated, we can have a **Deprecated** tab that says: "All versions less than v1.26.x are considered deprecated. Upgrade to a newer version to take advantage of new features."

- It could be we only need these tabs: **1.26.x** and **Deprecated**.

import Tooltip from "shared/components/ui/Tooltip";

# AWS Cluster Autoscaler 

Palette supports autoscaling for EKS clusters by using the AWS Cluster Autoscaler pack. 
The AWS Cluster Autoscaler dynamically scales cluster resources. It monitors the workload and provisions or shuts down cluster nodes to maximize cluster utilization. It resizes the Kubernetes cluster under the following conditions:
<br />

* Insufficient resources leading to multiple pod failures in the cluster. The AWS Cluster Autoscaler redistributes these pods to different nodes in such cases.

* Underutilized cluster nodes for a specific period. In this scenario, the AWS Cluster Autoscaler migrates the pods from underutilized nodes to other available nodes.

# Versions Supported

<Tabs>

<Tabs.TabPane tab="1.26.x" key="1.26.x">

## Prerequisites

* Kubernetes 1.24.x or higher. 

</Tabs.TabPane>

<Tabs.TabPane tab="1.24.x" key="1.24.x">

## Prerequisites

* Kubernetes 1.19.x or higher. 

</Tabs.TabPane>

<Tabs.TabPane tab="1.0.x" key="1.0.x">

## Prerequisites

* Kubernetes 1.19.x or higher. 

</Tabs.TabPane>

</Tabs>

* Permission to create an IAM policy in the AWS account you use with Palette. 

> * **IAM policy** - The following full Cluster Autoscaler policy should be attached to the cluster's node group. There are two ways to achieve this prerequisite. You can define this as a *customer-managed* policy in the AWS account and use its Amazon Resource Name (ARN) in the cluster profile. Alternatively, you can attach the following IAM policy as an *inline* policy to the node group if you have already deployed your cluster. Refer to the **Usage** section below for more details. <Tooltip>Refer to the [AWS IAM documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-vs-inline.html) for the differences between managed and inline policies.</Tooltip>

Reviewer 2: ```suggestion
*IAM policy - The following IAM policy must be attached to the EKS cluster's node group. There are two ways to achieve this prerequisite. You can define the policy as a *customer-managed* policy in the AWS account and use its Amazon Resource Name (ARN) in the cluster profile. Alternatively, you can attach the following IAM policy as an *inline* policy to the node group if you have already deployed your cluster. Refer to the **Usage** section below for more details. 
```

I removed the bold because it's not needed, and we also don't use tooltips. Link to the usage section.  You can link to an element inside the tabs, refer to [this](https://github.com/spectrocloud/librarium#tabs-component) section of the docs repo's README.

<br />

  ```json
  {
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Action": [
          "autoscaling:DescribeAutoScalingGroups",
          "autoscaling:DescribeAutoScalingInstances",
          "autoscaling:DescribeLaunchConfigurations",
          "autoscaling:DescribeScalingActivities",
          "autoscaling:DescribeTags",
          "ec2:DescribeInstanceTypes",
          "ec2:DescribeLaunchTemplateVersions"
        ],
        "Resource": ["*"]
      },
      {
        "Effect": "Allow",
        "Action": [
          "autoscaling:SetDesiredCapacity",
          "autoscaling:TerminateInstanceInAutoScalingGroup",
          "ec2:DescribeImages",
          "ec2:GetInstanceTypesFromInstanceRequirements",
          "eks:DescribeNodegroup"
        ],
        "Resource": ["*"]
      }
    ]
  }
  ```
<br />

> * **Updated Kubernetes layer manifest** - Kubernetes pack's manifest should be updated with the newly created IAM policy ARN. The YAML code block below shows the `managedMachinePool.roleAdditionalPolicies` section to update in the Kubernetes pack's manifest. Refer to the **Usage** section below for more details with an example. 

Reviewer 2: ```suggestion
* Updated Kubernetes layer manifest - The Kubernetes pack's manifest should be updated with the newly created IAM policy ARN. The YAML code block below shows the `managedMachinePool.roleAdditionalPolicies` section to update in the Kubernetes pack's manifest. Refer to the Usage section below for more details with an example. 
```

Removed the bold, and same feedback as above, link to the usage section.

<br />

  ```yaml
  managedMachinePool:
    #roleName: {{ name of the self-managed role | format "${string}" }}
    ## A list of additional policies to attach to the node group role
    roleAdditionalPolicies:
    - {{ arn for the policy1 | format "${string}" }}
  ```

## Usage

> The first step is to define an IAM policy in the AWS account associated with Palette. Then, switch to Palette to update the Kubernetes layer manifest and add the AWS Cluster Autoscaler pack. Finally, deploy your cluster and scale down the instance type to validate that AWS Cluster Autoscaler works correctly. The following sub-sections will provide more detailed instructions.

Reviewer 2: ```suggestion
To use the AWS Cluster Autoscaler pack, you must define an IAM policy in the AWS account associated with Palette. Then, switch to Palette to update the Kubernetes layer manifest and add the AWS Cluster Autoscaler pack.  The following sub-sections will provide more detailed instructions.
```

I removed the scale-down statement because it's not required unless the user wants to do that.

> The first step is to define an IAM policy in the AWS account associated with Palette. Then, switch to Palette to update the Kubernetes layer manifest and add the AWS Cluster Autoscaler pack. Finally, deploy your cluster and scale down the instance type to validate that AWS Cluster Autoscaler works correctly. The following sub-sections will provide more detailed instructions.

Reviewer 2: @sudkul87, for all these steps, number them so that the user understands that they need to be followed sequentially. See the Ubuntu pack page for example: https://docs.spectrocloud.com/integrations/ubuntu

<br />

### Define IAM Policy

> Define the new IAM policy, using the full Cluster Autoscaler policy outlined in the prerequisites section above, and give it a name, for example, *PaletteEKSClusterAutoscaler*. After defining the IAM policy, copy its ARN to use in the next step. For example, the policy ARN will be similar to `arn:aws:iam::650628870702:policy/PaletteEKSClusterAutoscaler`. 

Reviewer 2: ```suggestion
Define the new IAM policy, using the IAM policy outlined in the prerequisites section above, and give it a name, for example, *PaletteEKSClusterAutoscaler*. After defining the IAM policy, copy its ARN to use in the next step. For example, the policy ARN will be similar to `arn:aws:iam::650628870702:policy/PaletteEKSClusterAutoscaler`. 

```

![Snapshot showing the ARN  added to the Kubernetes pack's manifest.](/integrations_aws-cluster-autoscaler_k8s-manifest.png)

The code block below shows the updated `managedMachinePool.roleAdditionalPolicies` section with a sample policy ARN, `"arn:aws:iam::650628870702:policy/PaletteEKSClusterAutoscaler"`, as an example. Before you use the following code block, replace the ARN below with yours. 
<br />

  ```yaml
  managedMachinePool:
    # roleName: {{ name of the self-managed role | format "${string}" }}
    # A list of additional policies to attach to the node group role
    roleAdditionalPolicies:
    - "arn:aws:iam::650628870702:policy/PaletteEKSClusterAutoscaler"
  ```
  <br />

<InfoBox>

If you do not want to update the Kubernetes pack's manifest, you can add an *inline* IAM policy to the cluster's node group post deployment. Refer to this [AWS guide](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_manage-attach-detach.html#add-policies-console) on how to embed an inline policy for a user or role.

</InfoBox>

> Note that when you deploy an EKS cluster using Palette, Palette creates two IAM roles in the AWS account automatically - one for the cluster and another for the node group. Adding the ARN of your newly created IAM policy to the Kubernetes pack's manifest will make Palette add that IAM policy to the node group during deployment. 

Reviewer 2: ```suggestion
When you deploy an EKS cluster using Palette, Palette automatically creates two IAM roles in the AWS account. One for the cluster and another for the cluster's node group. Adding the ARN of your newly created IAM policy to the Kubernetes pack's manifest will make Palette add that IAM policy to the node group during deployment. 
```

Can we share the name of the roles, or the name prefix in case they have dynamic values at the end.

<br />

### Add Cluster Autoscaler Pack

> After updating your Kubernetes pack, add/update other layers per your requirements. Next, add AWS Cluster Autoscaler pack as an add-on layer. 

Reviewer 2: ```suggestion
After updating your Kubernetes pack, add or update other layers per your requirements. Next, add AWS Cluster Autoscaler pack as an add-on layer. 
```
We don't use `/` with sentences, use "or"

<br />

### Deploy a Cluster

> Once your cluster profile is complete, deploy it to a cluster. During the cluster configuration, ensure that your worker pool is made up of instance type **t3.2xlarge** (8 vCPUs, 32 GB RAM) or higher. As an example, the snapshot below shows one node of type **t3.2xlarge** in the worker pool of a successfully deployed cluster.  

Reviewer 2: This doesn't seem right. The reason I say this is because the autoscaler pack uses the [EC2 Cluster Autoscaler](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md) behind the scenes and that has no minimum reqs. What's more important is to convey to the user that they should ensure each instance size has adequate resources that provide approximately equal capacity. 

![CleanShot 2023-07-13 at 11 39 22](https://github.com/spectrocloud/librarium/assets/29551334/848f922a-7826-435a-a733-918dd88a7a9e)

AWS Cluster Autoscaler dynamically scales cluster nodes to maximize cluster utilization by monitoring the workload. If nodes are overutilized, it will provision new nodes to accommodate unscheduled pods. On the other hand, if nodes are underutilized, it will migrate the pods from underutilized to other available nodes and shut down underutilized ones.   

# Troubleshooting

1. If you are facing the `LimitExceeded: Cannot exceed quota for PoliciesPerRole:10` error in the cluster deployment logs, it may be because the default IAM role Palette creates for the node group already has 10 policies attached to it, and you are trying to attach one more. By default, your AWS account will have a quota of 10 managed policies per IAM role. To fix the error, follow the instruction on this [AWS guide](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html#reference_iam-quotas-entities) to request a quota increase. 

2. If you encounter an `executable aws-iam-authenticator not found` error in your terminal while trying to access your EKS cluster from your local machine, you can resolve it by installing the 

> [aws-iam-authenticator](https://docs.aws.amazon.com/eks/latest/userguide/install-aws-iam-authenticator.html) plugin on your local machine. 

Reviewer 2: Why are these two numbered? THat makes it seem like they are sequential steps. Let's remove the numbers.

import PointsOfInterest from 'shared/components/common/PointOfInterest';
import Tooltip from "shared/components/ui/Tooltip";

# AWS Cluster Autoscaler 

Palette supports autoscaling for EKS clusters by using the AWS Cluster Autoscaler pack. 
The Cluster Autoscaler dynamically scales cluster resources. It monitors the workload and provisions or shuts down cluster nodes to maximize cluster utilization. It resizes the Kubernetes cluster under the following conditions:
<br />

* Insufficient resources leading to multiple pod failures in the cluster. The Cluster Autoscaler redistributes these pods to different nodes in such cases.

* Underutilized cluster nodes for a specific period. In this scenario, the Cluster Autoscaler migrates the pods from underutilized nodes to other available nodes.

> Cluster Autoscaler pack runs as a `Deployment` in your cluster and utilizes [Amazon EC2 Auto Scaling Groups](https://docs.aws.amazon.com/autoscaling/ec2/userguide/what-is-amazon-ec2-auto-scaling.html) to manage node groups. 

Reviewer 2: ```suggestion
Cluster Autoscaler pack is deployed as a [*Deployment*](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) in your cluster and utilizes [Amazon EC2 Auto Scaling Groups](https://docs.aws.amazon.com/autoscaling/ec2/userguide/what-is-amazon-ec2-auto-scaling.html) to manage node groups. 
```

Run is an ableist word, so we avoid those. The *Deployment* is a reference to the Kubernetes object, Deployment, so we can italize it so that it stands out and not associated with the regular English word "Deployment."

# Versions Supported

<Tabs>

<Tabs.TabPane tab="1.26.x" key="1.26.x">

## Prerequisites

* Kubernetes 1.24.x or higher. 

</Tabs.TabPane>

<Tabs.TabPane tab="1.22.x" key="1.22.x">

## Prerequisites

* Kubernetes 1.19.x or higher. 

> </Tabs.TabPane>

Reviewer 2: @sudkul87 each tab gets its content. That means the Prerequisites, Parameters,  and Usage sections go inside of each version tab. This is explained in the [layout guide](https://spectrocloud.atlassian.net/wiki/spaces/DE/pages/1802797059/Packs).

  There are two ways to achieve this prerequisite. You can define the policy  as a *customer-managed* policy in the AWS account and use its Amazon Resource Name (ARN) in the cluster profile. Alternatively, you can attach the  IAM policy as an *inline* policy to the node group if you have already deployed your cluster. Refer to the [Usage](#usage) section below to learn more. 

* Updated Kubernetes layer manifest - The Kubernetes pack's manifest should be updated with the newly created IAM policy ARN. The YAML code block below shows the `managedMachinePool.roleAdditionalPolicies` section to update in the Kubernetes pack's manifest. Refer to the [Usage](#usage) section below for more details with an example. 
<br />

  ```yaml
  managedMachinePool:
    #roleName: {{ name of the self-managed role | format "${string}" }}
    ## A list of additional policies to attach to the node group role
    roleAdditionalPolicies:
    - {{ arn for the policy1 | format "${string}" }}
  ```
<br />

## Usage

> Cluster Autoscaler helps improve your cluster's performance and makes your cluster more resilient to failures. It automatically adjusts the number of nodes in your cluster based on the current workload. In other words, Cluster Autoscaler monitors the resource utilization, such as CPU and memory, and the number of pods running in your cluster and scales the cluster when either of these events occurs:

Reviewer 2: ```suggestion
Cluster Autoscaler helps improve your cluster's performance and makes your cluster more resilient to failures. It automatically adjusts the number of nodes in your cluster based on the current workload. In other words, Cluster Autoscaler monitors the resource utilization, such as CPU and memory, and the number of pods active in your cluster and scales the cluster when either of these events occurs:
```

Another ableism term

<br />

- Multiple pods fail due to resource contention. In this case, Cluster Autoscaler will provision more nodes. 

> - Nodes are underutilized for a specific period. In this case, Cluster Autoscaler will reschedule the pods onto other nodes and shut down the underutilized node. 

Reviewer 2: ```suggestion
- Nodes are underutilized for a specific period. In this case, the Cluster Autoscaler will reschedule the pods onto other nodes and shut down the underutilized node. 

```

> To know the internal details of the pack, check out the pack manifest [here](https://github.com/spectrocloud/pax/tree/master/stable/addon/systemapps). If you want to customize the pack to use custom metrics, you must modify the existing pack and follow the instructions in the [Create and Deploy a Custom Add-On Pack](/registries-and-packs/deploy-pack) guide to deploy your custom pack. 

Reviewer 2: ```suggestion
```

Best to remove this as users don't have access to the pax repository.

<br />

  ```yaml
  managedMachinePool:
    # roleName: {{ name of the self-managed role | format "${string}" }}
    # A list of additional policies to attach to the node group role
    roleAdditionalPolicies:
    - "arn:aws:iam::650628870702:policy/PaletteEKSClusterAutoscaler"
  ```
  <br />

  <InfoBox>

  If you do not want to update the Kubernetes pack's manifest, you can add an *inline* IAM policy to the cluster's node group post deployment. Refer to this [AWS guide](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_manage-attach-detach.html#add-policies-console) on how to embed an inline policy for a user or role. Refer to the [AWS IAM documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-vs-inline.html) for the differences between managed and inline policies.

  </InfoBox>

4. In the cluster deployment wizard, when you are in the **Nodes configuration** section, you must enter the minimum and maximum number of worker pool nodes, and the instance type (size) that suits your requirement.

>   You need to enter these node count limits because Cluster Autoscaler uses Auto Scaling Group to manage the cluster's node group. Auto Scaling Group requires a minimum and maximum count and the instance type. Also, specify the instance type (size). You can choose an instance type that suits your requirement. 

Reviewer 2: ```suggestion
  You must provide the node count limits because the Cluster Autoscaler uses an Auto Scaling Group to manage the cluster's node group. An Auto Scaling Group requires a minimum and maximum count and the selection of an instance type.  You can choose an instance type that suits your requirement. 
```
Minor touch-up. There was some redundancy.

>   For example, the snapshot below shows the minimum one and maximum three node count for the Cluster Autoscaler. 

Reviewer 2: ```suggestion
  For example, the snapshot below displays the cluster's minimum and maximum capacity.
```

A more concise way of writing the statement.

  ![A snapshot showing the minimum and maximum node count in Palette.](/integrations_aws-cluster-autoscaler_node-count.png)

<br />

### Resize the Cluster

> This section illustrates a use case to bring the Cluster Autoscaler into action. In this example, you will first create a cluster with large-sized worker pool instances. Next, you will manually reduce the instance size, leading to insufficient resources for existing pods and multiple pod failures in the cluster. As a result, Cluster Autoscaler will provision new smaller-sized nodes with enough capacity to accommodate the current workload and reschedule those contending pods on new nodes. Also, the new nodes' count will be within the minimum and maximum limit you specified for the worker pool.

Reviewer 2: ```suggestion
To better understand the scaling behavior of the Cluster Autoscaler and its impact on a cluster, do the following exercise to gain firsthand experience with the scaling behavior. 

In the following example scenario, you will first create a cluster with large-sized worker pool instances. Next, you will manually reduce the instance size, leading to insufficient resources for existing pods and multiple pod failures in the cluster. As a result, the Cluster Autoscaler will provision new smaller-sized nodes with enough capacity to accommodate the current workload and reschedule those contending pods on new nodes. Also, the new nodes' count will be within the minimum and maximum limit you specified for the worker pool.
```

I touched up the intro to make it sound more natural and provide the reader an understanding that it's a learning exercise.

> For the current example, use the following steps to trigger the pod rescheduling event manually:

Reviewer 2: ```suggestion
Use the following steps to trigger the pod rescheduling event manually:
```
This can be a more concise sentence.

<br />

1. In the cluster deployment wizard, while defining the **Nodes configuration**, choose a large-sized instance type. For example, you can choose your worker pool to have instance size **t3.2xlarge** (8 vCPUs, 32 GB RAM) or higher. 

> 2. After your cluster is successfully deployed, navigate to the **Nodes** tab in the cluster details page in Palette, and note the count and size of nodes. For example, the snapshot below shows one node of type **t3.2xlarge** in the worker pool of a successfully deployed cluster.   

Reviewer 2: ```suggestion
2. After your cluster is successfully deployed, navigate to the **Nodes** tab in the cluster details page in Palette, and note the count and size of nodes. The snapshots below display one node of the type **t3.2xlarge** in the worker pool of a successfully deployed cluster.   

```

  ![A snapshot showing one node of type **t3.2xlarge** in the worker pool.](/integrations_aws-cluster-autoscaler_one-node.png)

> 3. Manually reduce the instance size in the worker-pool configuration. For example, reduce the instance size to **t3.medium** (2 vCPUs, 8 GB RAM). The snapshot below shows how to edit the instance size in the node pool configuration.

Reviewer 2: ```suggestion
3. Manually reduce the instance size in the worker-pool configuration to a **t3.medium** (2 vCPUs, 8 GB RAM). The snapshot below shows how to edit the instance size in the node pool configuration.
```

This is your example scenario. You can just state what size to reduce it to.

  ![A snapshot showing how to edit node pool configuration.](/integrations_aws-cluster-autoscaler_edit-node.png)

> 4. Wait for a few minutes for the new nodes to provision. Reducing the node size will make the Cluster Autoscaler shut down the large node and provision smaller-sized nodes with enough capacity to accommodate the current workload. Also, the new nodes' count will be within the minimum and maximum limit you specified for the worker pool. 

Reviewer 2: ```suggestion
4. Wait for a few minutes for the new nodes to provision. Reducing the node size will make the Cluster Autoscaler shut down the large node and provision smaller-sized nodes with enough capacity to accommodate the current workload. Also, the new node count will be within the minimum and maximum limit you specified for the worker pool configuration wizard. 

```

>   For example, the snapshot below shows two new nodes of size **t3.medium** spin up automatically. These two smaller-sized nodes will be able to handle the workload just as well as the single large-sized node. 

Reviewer 2: ```suggestion
The following snapshot displays two new nodes of the size **t3.medium** spin up automatically. These two smaller-sized nodes will be able to handle the same workload as a single larger-sized node. 

```

>   ![A snapshot showing new nodes of size **t3.medium** spin up automatically, *collectively* providing enough capacity to accommodate the current workload. ](/integrations_aws-cluster-autoscaler_two-nodes.png)

Reviewer 2: ```suggestion
  ![A snapshot displaying new nodes of the size **t3.medium** spin up automatically, and *collectively* providing enough capacity to accommodate the current workload. ](/integrations_aws-cluster-autoscaler_two-nodes.png)

```

  <br />

# Troubleshooting

> If you are facing the `LimitExceeded: Cannot exceed quota for PoliciesPerRole:10` error in the cluster deployment logs, it may be because the default IAM role Palette creates for the node group already has 10 policies attached to it, and you are trying to attach one more. By default, your AWS account will have a quota of 10 managed policies per IAM role. To fix the error, follow the instruction on this [AWS guide](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html#reference_iam-quotas-entities) to request a quota increase. 

Reviewer 2: ```suggestion
If you are facing the `LimitExceeded: Cannot exceed quota for PoliciesPerRole:10` error in the cluster deployment logs, it may be because the default IAM role Palette creates for the node group already has 10 policies attached to it, and you are trying to attach one more. By default, your AWS account will have a quota of 10 managed policies per IAM role. To fix the error, follow the instruction in this [AWS guide](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html#reference_iam-quotas-entities) to request a quota increase. 

```

> If you encounter an `executable aws-iam-authenticator not found` error in your terminal while trying to access your EKS cluster from your local machine, you can resolve it by installing the 
> [aws-iam-authenticator](https://docs.aws.amazon.com/eks/latest/userguide/install-aws-iam-authenticator.html) plugin on your local machine. 

Reviewer 2: ```suggestion
If you encounter an `executable aws-iam-authenticator not found` error in your terminal when attempting to access your EKS cluster from your local machine, it may be due to the [aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator) plugin missing from your local environment. You can find the installation steps for the 
aws-iam-authenticator in the following [install guide](https://docs.aws.amazon.com/eks/latest/userguide/install-aws-iam-authenticator.html). 
```

The current form of the paragraph needed to be broken up to clarify the issue and how to resolve it.

# References

- [Cluster Autoscaler on AWS](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md)

> - [Amazon EKS Autoscaling](https://docs.aws.amazon.com/eks/latest/userguide/autoscaling.html)

Reviewer 2: ```suggestion
- [Amazon EKS Autoscaling](https://docs.aws.amazon.com/eks/latest/userguide/autoscaling.html)

- [AWS IAM Authenticator Plugin](https://github.com/kubernetes-sigs/aws-iam-authenticator)
```

# AWS Cluster Autoscaler

> The AWS Cluster Autoscaler dynamically scales cluster resources. It monitors the workload and provisions or shuts down cluster nodes to maintain the desired cluster utilization. It resizes the Kubernetes cluster under the following conditions:

Reviewer 2: I'm pretty sure this is only applicable to EKS, so we should probably mention that.

<br />

* Insufficient resources leading to multiple pod failures in the cluster. The AWS Cluster Autoscaler redistributes these pods to different nodes in such cases.

* Underutilized cluster nodes for a specific period. In this scenario, the AWS Cluster Autoscaler migrates the pods from underutilized nodes to other available nodes.

# Versions Supported

<Tabs>

<Tabs.TabPane tab="1.26.x" key="1.26.x">

## Prerequisite

> * Kubernetes server (control plane) version 1.24.x or higher. 

Reviewer 2: ```suggestion
* Kubernetes 1.24.x or higher. 
```
Keep it simple

          "eks:DescribeNodegroup"
        ],
        "Resource": ["*"]
      }
    ]
  }
  ```

* **Kubernetes Layer Configuration** - Update the Kubernetes pack's YAML configuration. In the `managedMachinePool` node group, add the Amazon Resource Name (ARN) of the IAM policy created for the service account. The exmaple below shows how you can add the IAM policy's ARN to your Kubernetes pack's YAML configuration. In this example, replace the `"arn:aws:iam::012345678910:policy/autoscalingpolicy"` with the ARN of your IAM policy. 
<br />

  ```yaml
  managedMachinePool:
    # roleName: {{ name of the self-managed role | format "${string}" }}
    # A list of additional policies to attach to the node group role
    roleAdditionalPolicies:
    - "arn:aws:iam::012345678910:policy/autoscalingpolicy"
  ```

## Usage

Deploy a AWS EKS cluster with the AWS Cluster Autoscaler pack. Ensure that your worker pool is made up of instance size **t3.2xlarge** (8 vCPUs, 32 GB RAM) or higher, and then manually scale the instance size down to **t3.medium** (2 vCPUs, 8 GB RAM). Scaling down the instance size will result in insufficient resources for some pods. 

> Suppose you have configured the AWS Cluster Autoscaler correctly. In that case, it will recognize the node's unscheduled pods and scale up the cluster by creating an additional node to accommodate them.

Reviewer 2: I would advice you go play with the pack and attempt to consume the pack. Experience for yourself the behavior of scaling the cluster up and down. That way you can better explain the behavior.  Snehal has a demo in [Eng Weekly](https://spectrocloud.atlassian.net/wiki/spaces/ENGINEERIN/pages/1726775297/Eng+Demos)that  you can check out for Dec 7, 2022.

# AWS Cluster Autoscaler

Palette supports autoscaling for AWS EKS clusters, ensuring cluster nodes launch according to the workload. 

The AWS cluster autoscaler dynamically scales resources up or down based on the changing cluster workload. It monitors the workload and provisions or shuts down cluster nodes to maintain consistent cluster utilization. It resizes the Kubernetes cluster under the following conditions:

* Insufficient resources leading to multiple pod failures in the cluster. The AWS cluster autoscaler redistributes these pods to different nodes in such cases.

* Underutilized cluster nodes for a specific period. In this scenario, the AWS cluster autoscaler migrates the pods from underutilized nodes to other available nodes.

# Versions Supported

<Tabs>

<Tabs.TabPane tab="1.24+" key="1.24.x">

> **1.24+**

Reviewer: @sudkul87 Delete line 33. Our packs layout format doesn't list each version.
Packs Layout: https://spectrocloud.atlassian.net/wiki/spaces/DE/pages/1800994817/Packs+Layout

> **1.24+**

Reviewer 3: The versions supported are `1.22.2` and `1.26.3` . 1.24 is the k8s version that supports these packs

> You can use AWS cluster autoscaler with the Kubernetes control plane (previously referred to as *master*) version 1.24.x or higher. 

Reviewer: Can we just say:

"You can use the AWS Cluster Autoscaler pack with the Kubernetes master node." 

Are we using two terms for the same thing? "control plane" and "master node"? I think the UI uses the term "master node".

# Prerequisite

> * Kubernetes version 1.24.x and above.

Reviewer: This may not be a prerequisite. Instead, we show the supported Kubernetes versions as tabs.

          "autoscaling:DescribeAutoScalingGroups",
          "autoscaling:DescribeAutoScalingInstances",
          "autoscaling:DescribeLaunchConfigurations",
          "autoscaling:DescribeScalingActivities",
          "autoscaling:DescribeTags",
          "ec2:DescribeInstanceTypes",
          "ec2:DescribeLaunchTemplateVersions"
        ],
        "Resource": ["*"]
      },
      {
        "Effect": "Allow",
        "Action": [
          "autoscaling:SetDesiredCapacity",
          "autoscaling:TerminateInstanceInAutoScalingGroup",
          "ec2:DescribeImages",
          "ec2:GetInstanceTypesFromInstanceRequirements",
          "eks:DescribeNodegroup"
        ],
        "Resource": ["*"]
      }
    ]
  }
  ```

* Update the Kubernetes pack's  `managedMachinePool` node group with the **ARN** of the autoscaler policy created for the service account. Here is an example YAML configuration. 
<br />

  ```yaml
  managedMachinePool:
    #roleName: {{ name of the self-managed role | format "${string}" }}

    ## A list of additional policies to attach to the node group role
    roleAdditionalPolicies:
    - "arn:aws:iam::012345678910:policy/autoscalingpolicy"
  ```

> ## Usage

Reviewer: This Usage section might be all you need for this pack, unless Engineering has additional information.

Deploy an EKS cluster with a cluster autoscaler pack. Ensure that your worker pool is made up of instance size **t3.2xlarge** (8 vCPUs, 32 GB RAM) or higher, and then manually scale the instance size down to **t3.medium** (2 vCPUs, 8 GB RAM). Scaling down the instance size will result in insufficient resources for some pods. 

Suppose you have configured the AWS cluster autoscaler correctly. In that case, it will recognize the node's unscheduled pods and scale up the cluster by creating an additional node to accommodate them.

> To confirm the creation of the new node, review the pod logs.

Reviewer: ```suggestion
To confirm the creation of the new node, you can review the pod logs.

```
